pkg/kubectl/polymorphichelpers/mapbasedselectorforobject.go

```diff
@@ -50,27 +50,52 @@ func mapBasedSelectorForObject(object runtime.Object) (string, error) {
 		return generate.MakeLabels(t.Spec.Selector), nil
 
 	case *extensionsv1beta1.Deployment:
-		// TODO(madhusudancs): Make this smarter by admitting MatchExpressions with Equals
-		// operator, DoubleEquals operator and In operator with only one element in the set.
-		if len(t.Spec.Selector.MatchExpressions) > 0 {
-			return "", fmt.Errorf("couldn't convert expressions - \"%+v\" to map-based selector format", t.Spec.Selector.MatchExpressions)
+		// "extensions" deployments use pod template labels if selector is not set.
+		var labels map[string]string
+		if t.Spec.Selector != nil {
+			// TODO(madhusudancs): Make this smarter by admitting MatchExpressions with Equals
+			// operator, DoubleEquals operator and In operator with only one element in the set.
+			if len(t.Spec.Selector.MatchExpressions) > 0 {
+				return "", fmt.Errorf("couldn't convert expressions - \"%+v\" to map-based selector format", t.Spec.Selector.MatchExpressions)
+			}
+			labels = t.Spec.Selector.MatchLabels
+		} else {
+			labels = t.Spec.Template.Labels
 		}
-		return generate.MakeLabels(t.Spec.Selector.MatchLabels), nil
+		if len(labels) == 0 {
+			return "", fmt.Errorf("the deployment has no labels or selectors and cannot be exposed")
+		}
+		return generate.MakeLabels(labels), nil
+
 	case *appsv1.Deployment:
```

Comment:
All the `apps` cases are the same. Combine them into one?
```go 
case *appsv1.Deployment, *appsv1beta2.Deployment, *appsv1beta1.Deployment:
```

Reply:
It appears that type switches can not be combined if the type is used in the case statement; it will not compile. The compilation error is:

t.Spec undefined (type "k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime".Object has no field or method Spec)

The compiler can not disambiguate the "t.Spec", since "t" is three different types.